src/NITTA/Model/Microarchitecture/Builder.hs

```diff
@@ -42,26 +43,33 @@ data BuilderSt tag v x t = BuilderSt
     { signalBusWidth :: Int
     , availPorts :: [SignalTag]
     , puProtos :: [(tag, PU v x t)]
-    , netEnv :: UnitEnv (BusNetwork tag v x t)
     }
 
+netIOPorts puProtos =
+    let setOfPorts ports = foldl (flip $ S.union . ports . snd) S.empty puProtos
+     in BusNetworkIO
+            { extInputs = setOfPorts puInputPorts
+            , extOutputs = setOfPorts puOutputPorts
+            , extInOuts = setOfPorts puInOutPorts
+            }
+
+puEnv tag ports ioports =
+    def
+        { ctrlPorts = Just ports
+        , ioPorts = Just ioports
+        , valueIn = Just ("data_bus", "attr_bus")
+        , valueOut = Just (toText tag <> "_data_out", toText tag <> "_attr_out")
+        }
+
 -- |Define microarchitecture with BusNetwork
 defineNetwork bnName ioSync builder =
-    let defEnv = def
-        st0 =
+    let st0 =
             BuilderSt
                 { signalBusWidth = 0
                 , availPorts = map (SignalTag . controlSignalLiteral) [0 :: Int ..]
                 , puProtos = []
-                , netEnv = defEnv
-                }
-        BuilderSt{signalBusWidth, puProtos} = flip execState st0 $ void builder
-        netIOPorts =
-            BusNetworkIO
-                { extInputs = L.nub $ concatMap (puInputPorts . snd) puProtos
-                , extOutputs = L.nub $ concatMap (puOutputPorts . snd) puProtos
-                , extInOuts = L.nub $ concatMap (puInOutPorts . snd) puProtos
                 }
+        BuilderSt{signalBusWidth, puProtos, availPorts} = flip execState st0 $ void builder
```

Comment:
I don’t feel that void here is needed. Because without it, you move to restrict builder type to return nothing, which decreases the chance of mistake of user.